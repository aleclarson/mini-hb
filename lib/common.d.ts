import { Context } from './context';
export declare type AnyObj = {
    [key: string]: any;
};
export declare type RenderFn = (template: string, ...contexts: AnyObj[]) => string;
export declare type TemplateFn = (...contexts: AnyObj[]) => string;
export interface IHandleBars {
    bind: IBindFn;
    /** Parse a template and render it into a string */
    (template: string, ...contexts: AnyObj[]): string;
    /** Create an immutable context that can be reused */
    (...contexts: AnyObj[]): Context;
}
export interface IBoundProps {
    readonly template?: string;
    readonly context: Context;
}
export interface IBindFn extends Function {
    /** Bind a template (and optionally some context) to the renderer */
    (template: string, ...contexts: AnyObj[]): TemplateFn & IBoundProps;
    /** Bind some context to the renderer */
    (...contexts: AnyObj[]): RenderFn & IBoundProps;
}
export declare const isArray: (arg: any) => arg is any[];
/** Deeply flatten an array of arrays */
export declare function flatMap<T, U>(src: T[], mapFn: (val: T, i: number) => U, dest?: U[]): U[];
export declare function dlv(obj: AnyObj, path: string[], def?: any): any;
