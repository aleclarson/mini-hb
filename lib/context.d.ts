import { AnyObj } from './common';
/** A stack of variable maps */
export declare class Context {
    readonly stack: AnyObj[];
    constructor(stack?: AnyObj[]);
    get(key: string): any;
    concat(...contexts: AnyObj[]): Context;
}
