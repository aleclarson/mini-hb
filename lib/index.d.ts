import { IHandleBars } from './common';
export declare const hb: IHandleBars;
