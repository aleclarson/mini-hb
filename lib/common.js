"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.isArray = Array.isArray;
/** Deeply flatten an array of arrays */
function flatMap(src, mapFn, dest) {
    if (dest === void 0) { dest = []; }
    for (var i = 0, val = void 0; i < src.length; i++) {
        exports.isArray((val = src[i]))
            ? flatMap(val, mapFn, dest)
            : dest.push(mapFn(val, i));
    }
    return dest;
}
exports.flatMap = flatMap;
// https://github.com/developit/dlv
function dlv(obj, key, def) {
    var p = 0;
    key = exports.isArray(key) ? key : key.split('.');
    while (obj && p < key.length)
        obj = obj[key[p++]];
    return obj === undefined || p < key.length ? def : obj;
}
exports.dlv = dlv;
