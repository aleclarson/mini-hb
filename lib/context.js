Object.defineProperty(exports, "__esModule", { value: true });
var common_1 = require("./common");
var unwrapContext = function (ctx) {
    return ctx instanceof Context ? ctx.stack : ctx;
};
/** A stack of variable maps */
var Context = /** @class */ (function () {
    function Context(stack) {
        this.stack = stack ? common_1.flatMap(stack, unwrapContext) : [];
    }
    Context.prototype.get = function (key) {
        var stack = this.stack;
        for (var i = stack.length; i > 0;) {
            var value = common_1.dlv(stack[--i], key);
            if (value !== undefined) {
                return value;
            }
        }
        return '';
    };
    Context.prototype.concat = function () {
        var contexts = [];
        for (var _i = 0; _i < arguments.length; _i++) {
            contexts[_i] = arguments[_i];
        }
        var stack = this.stack.concat(common_1.flatMap(contexts, unwrapContext));
        return stack.length == this.stack.length ? this : new Context(stack);
    };
    return Context;
}());
exports.Context = Context;
