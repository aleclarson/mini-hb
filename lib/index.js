"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var context_1 = require("./context");
var parser_1 = require("./parser");
var render_1 = require("./render");
exports.hb = (function ($0) {
    var contexts = [];
    for (var _i = 1; _i < arguments.length; _i++) {
        contexts[_i - 1] = arguments[_i];
    }
    if (typeof $0 == 'string') {
        var ctx = contexts.length == 1 && contexts[0] instanceof context_1.Context
            ? contexts[0]
            : new context_1.Context(contexts);
        return render_1.render($0, parser_1.parse($0), ctx);
    }
    else {
        return $0 instanceof context_1.Context
            ? $0.concat(contexts)
            : (contexts.unshift($0), new context_1.Context(contexts));
    }
});
// The `bind` method
exports.hb.bind = function ($0) {
    var contexts = [];
    for (var _i = 1; _i < arguments.length; _i++) {
        contexts[_i - 1] = arguments[_i];
    }
    var bound;
    var ctx = new context_1.Context(contexts);
    if (typeof $0 == 'string') {
        var template_1 = $0;
        var nodes_1 = parser_1.parse($0);
        bound = function () {
            var contexts = [];
            for (var _i = 0; _i < arguments.length; _i++) {
                contexts[_i] = arguments[_i];
            }
            return render_1.render(template_1, nodes_1, ctx.concat(contexts));
        };
        bound.template = template_1;
    }
    else {
        ctx.stack.unshift($0);
        bound = function (template) {
            var contexts = [];
            for (var _i = 1; _i < arguments.length; _i++) {
                contexts[_i - 1] = arguments[_i];
            }
            return render_1.render(template, parser_1.parse(template), ctx.concat(contexts));
        };
    }
    bound.context = ctx;
    return bound;
};
