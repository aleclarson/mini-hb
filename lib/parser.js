Object.defineProperty(exports, "__esModule", { value: true });
// Character codes
var LINE_BREAK = '\n'.charCodeAt(0);
var L_BRACKET = '{'.charCodeAt(0);
var R_BRACKET = '}'.charCodeAt(0);
var COMMA = ','.charCodeAt(0);
var SPACE = ' '.charCodeAt(0);
var SLASH = '/'.charCodeAt(0);
var HASH = '#'.charCodeAt(0);
/** For validating variable names */
var VAR_RE = /^[a-z_$][a-z0-9_.$]*$/i;
/** For removing the first level of indentation */
var INDENT_RE = /(^|\n)  /g;
/** The first line (if empty) */
var FIRST_LINE_EMPTY = /^[ ]*\n/;
/** The last line (if empty) */
var LAST_LINE_EMPTY = /\n[ ]*$/;
/** For defining a non-nullable property. */
var nil = undefined;
/** @internal */
function parse(template) {
    /** List of top-level nodes */
    var nodes = [];
    /** 0-based char offset */
    var i = -1;
    /** 1-based line offset */
    var line = 1;
    /** Current in-progress node */
    var node = null;
    /** Current block */
    var block = null;
    while (i < template.length - 1) {
        var ch = template.charCodeAt(++i);
        if (ch == LINE_BREAK) {
            line++;
            continue;
        }
        if (!node) {
            // Parse the opening brackets.
            if (ch !== L_BRACKET)
                continue;
            ch = template.charCodeAt(i + 1);
            if (ch !== L_BRACKET)
                continue;
            i++;
            // Create an unclosed node.
            node = {
                type: nil,
                name: nil,
                args: nil,
                start: i - 1,
                line: line,
                end: nil,
            };
            continue;
        }
        // Parse the closing brackets.
        if (ch !== R_BRACKET)
            continue;
        ch = template.charCodeAt(i + 1);
        if (ch !== R_BRACKET)
            continue;
        i++;
        // Extract the contents between {{ and }}
        var content = template.slice(node.start + 2, i - 1).trim();
        node.end = i + 1;
        // The first char indicates the node type.
        ch = content.charCodeAt(0);
        // Opening a block.
        if (ch == HASH) {
            node.type = '#';
            block = {
                head: parseNode(node, content, !!block),
                body: nil,
                tail: nil,
                parent: block,
            };
        }
        // Closing a block.
        else if (ch == SLASH) {
            node.type = '/';
            parseNode(node, content, !!block);
            if (!block || node.name !== block.head.name) {
                var err = new SyntaxError("Unexpected closing node on line " + node.line);
                err.node = node;
                throw err;
            }
            if (!block.parent) {
                var body = template.slice(block.head.end, node.start);
                block.body =
                    node.line == block.head.line
                        ? body
                        : body
                            .replace(FIRST_LINE_EMPTY, '') // remove first line (if empty)
                            .replace(LAST_LINE_EMPTY, '') // remove last line (if empty)
                            .replace(INDENT_RE, '$1'); // dedent one level
                block.tail = node;
                nodes.push(block);
            }
            block = block.parent;
        }
        // Push top-level nodes only.
        else if (!block) {
            nodes.push(parseNode(node, content));
        }
        // The node is fully parsed.
        node = null;
    }
    // Were all nodes closed?
    if (node) {
        var err = new SyntaxError("Unclosed node on line " + node.line);
        err.node = node;
        throw err;
    }
    // Were all blocks closed?
    if (block) {
        var err = new SyntaxError("Unclosed block on line " + block.head.line);
        err.node = block.head;
        throw err;
    }
    return nodes;
}
exports.parse = parse;
/** For parsing the content between `{{` and `}}` */
function parseNode(node, content, skipArgs) {
    var i = node.type ? 1 : 0;
    // Parse the variable/function name.
    var name = '';
    for (; i < content.length; i++) {
        if (content.charCodeAt(i) == SPACE)
            break;
        name += content[i];
    }
    // Validate the name.
    if (VAR_RE.test(name)) {
        node.name = name;
    }
    else {
        throw badVariableName(name, node);
    }
    // Parse the arguments.
    if (!skipArgs && ++i < content.length) {
        var arg = '';
        var args = [];
        for (; i < content.length; i++) {
            var ch = content.charCodeAt(i);
            if (ch == SPACE)
                continue;
            if (ch == COMMA) {
                args.push(arg);
                arg = '';
            }
            else {
                arg += content[i];
            }
        }
        args.push(arg);
        node.args = args;
    }
    return node;
}
function badVariableName(name, node) {
    var err = new SyntaxError("Invalid variable name \"" + name + "\" on line " + node.line);
    err.node = node;
    return err;
}
