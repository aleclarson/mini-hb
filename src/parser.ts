import { AnyObj } from './common'
import { Context } from './context'

// Character codes
const LINE_BREAK = '\n'.charCodeAt(0)
const L_BRACKET = '{'.charCodeAt(0)
const R_BRACKET = '}'.charCodeAt(0)
const COMMA = ','.charCodeAt(0)
const SPACE = ' '.charCodeAt(0)
const SLASH = '/'.charCodeAt(0)
const HASH = '#'.charCodeAt(0)

/** For validating variable names */
const VAR_RE = /^[a-z_$][a-z0-9_.$]*$/i

/** For removing the first level of indentation */
const INDENT_RE = /(^|\n)  /g

/** The first line (if empty) */
const FIRST_LINE_EMPTY = /^[ ]*\n/

/** The last line (if empty) */
const LAST_LINE_EMPTY = /\n[ ]*$/

/** For defining a non-nullable property. */
const nil = undefined as any

type Options = {
  context?: AnyObj | Context
}

/** Each node represents a pair of `{{` and `}}` */
export interface Node {
  /** For blocks, this equals "#" for the head and "/" for the tail */
  type?: '#' | '/'
  /** The reference name */
  name: string
  /** Arguments to a function call */
  args?: string[]
  /** Character offset where this node begins */
  start: number
  /** The line number this node is on */
  line: number
  /** Character offset where this node ends */
  end: number
}

/** Each block represents a pair of `{{ #foo }}` and `{{ /foo }}` */
export interface Block {
  head: Node
  body: string
  tail: Node
  parent: Block | null
}

/** For syntax errors thrown by the parser */
export interface IParserError extends SyntaxError {
  node: Node
}

/** @internal */
export function parse(template: string) {
  /** List of top-level nodes */
  let nodes: Array<Node | Block> = []

  /** 0-based char offset */
  let i = -1

  /** 1-based line offset */
  let line = 1

  /** Current in-progress node */
  let node: Node | null = null

  /** Current block */
  let block: Block | null = null

  while (i < template.length - 1) {
    let ch = template.charCodeAt(++i)
    if (ch == LINE_BREAK) {
      line++
      continue
    }

    if (!node) {
      // Parse the opening brackets.
      if (ch !== L_BRACKET) continue
      ch = template.charCodeAt(i + 1)
      if (ch !== L_BRACKET) continue
      i++

      // Create an unclosed node.
      node = {
        type: nil,
        name: nil,
        args: nil,
        start: i - 1,
        line,
        end: nil,
      }
      continue
    }

    // Parse the closing brackets.
    if (ch !== R_BRACKET) continue
    ch = template.charCodeAt(i + 1)
    if (ch !== R_BRACKET) continue
    i++

    // Extract the contents between {{ and }}
    let content = template.slice(node.start + 2, i - 1).trim()
    node.end = i + 1

    // The first char indicates the node type.
    ch = content.charCodeAt(0)

    // Opening a block.
    if (ch == HASH) {
      node.type = '#'
      block = {
        head: parseNode(node, content, !!block),
        body: nil,
        tail: nil,
        parent: block,
      }
    }

    // Closing a block.
    else if (ch == SLASH) {
      node.type = '/'
      parseNode(node, content, !!block)

      if (!block || node.name !== block.head.name) {
        let err = new SyntaxError(
          `Unexpected closing node on line ${node.line}`
        ) as IParserError
        err.node = node
        throw err
      }

      if (!block.parent) {
        let body = template.slice(block.head.end, node.start)
        block.body =
          node.line == block.head.line
            ? body
            : body
                .replace(FIRST_LINE_EMPTY, '') // remove first line (if empty)
                .replace(LAST_LINE_EMPTY, '') // remove last line (if empty)
                .replace(INDENT_RE, '$1') // dedent one level

        block.tail = node
        nodes.push(block)
      }
      block = block.parent
    }

    // Push top-level nodes only.
    else if (!block) {
      nodes.push(parseNode(node, content))
    }

    // The node is fully parsed.
    node = null
  }

  // Were all nodes closed?
  if (node) {
    let err = new SyntaxError(
      `Unclosed node on line ${node.line}`
    ) as IParserError
    err.node = node
    throw err
  }

  // Were all blocks closed?
  if (block) {
    let err = new SyntaxError(
      `Unclosed block on line ${block.head.line}`
    ) as IParserError
    err.node = block.head
    throw err
  }

  return nodes
}

/** For parsing the content between `{{` and `}}` */
function parseNode(node: Node, content: string, skipArgs?: boolean) {
  let i = node.type ? 1 : 0

  // Parse the variable/function name.
  let name = ''
  for (; i < content.length; i++) {
    if (content.charCodeAt(i) == SPACE) break
    name += content[i]
  }

  // Validate the name.
  if (VAR_RE.test(name)) {
    node.name = name
  } else {
    throw badVariableName(name, node)
  }

  // Parse the arguments.
  if (!skipArgs && ++i < content.length) {
    let arg = ''
    let args = []
    for (; i < content.length; i++) {
      let ch = content.charCodeAt(i)
      if (ch == SPACE) continue
      if (ch == COMMA) {
        args.push(arg)
        arg = ''
      } else {
        arg += content[i]
      }
    }
    args.push(arg)
    node.args = args
  }

  return node
}

function badVariableName(name: string, node: Node) {
  let err = new SyntaxError(
    `Invalid variable name "${name}" on line ${node.line}`
  ) as IParserError
  err.node = node
  return err
}
