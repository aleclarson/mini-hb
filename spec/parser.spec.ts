import { parse } from '../lib/parser'

const parseSnapshot = (template: string) => {
  let nodes = parse(template)
  expect(nodes).toMatchSnapshot()
  return nodes
}

const parseThrows = (template: string) =>
  expect(() => parse(template)).toThrowErrorMatchingSnapshot()

test('empty node', () => {
  parseThrows('{{}}')
})

test('line breaks', () => {
  parseSnapshot(' \n \n{{a}}\n\n {{b}} \n')
})

describe('variable node', () => {
  test('with common padding', () => {
    parseSnapshot('{{a}}')
    parseSnapshot('{{ a }}')
  })
  test('with creative padding', () => {
    parseSnapshot('{{ a}}')
    parseSnapshot('{{  a    }}')
  })
  test('with spacing around', () => {
    parseSnapshot('  {{a }} ')
  })
  test('that spans multiple lines', () => {
    parseSnapshot(`
      foo
      {{
        a
      }}
      bar
    `)
  })
})

describe('variable names', () => {
  test('follow Javascript rules exactly', () => {
    parseSnapshot('{{ __aAoOzZ_$0123456789$$ }}')
  })
  test('can use dot-notation', () => {
    parseSnapshot('{{ a.b.c.d }}')
  })
  test('cannot use bracket-notation (yet?)', () => {
    parseThrows('{{ a[b[c]][d] }}')
  })
  test('cannot be a number', () => {
    parseThrows('{{ 12345 }}')
    parseThrows('{{ 12.34 }}')
  })
  test('cannot start with a number', () => {
    parseThrows('{{ 0z }}')
  })
  test('cannot contain dashes', () => {
    parseThrows('{{ a--b-c }}')
  })
})

describe('function call', () => {
  test('with one argument', () => {
    parseSnapshot('{{ fn z }}')
  })
  test('with many arguments', () => {
    let args = new Array(10)
      .fill(null)
      .map((_, i) => 'z' + i)
      .join(', ')

    parseSnapshot(`{{ fn ${args} }}`)
  })
  test('with primitive arguments', () => {
    parseSnapshot('{{ fn 0, -0.5, 1, true, false, null }}')
  })
  test('with no argument spacing', () => {
    parseSnapshot('{{fn 1,2,a,b,null}}')
  })
  test('with creative spacing', () => {
    parseSnapshot('{{ fn  1, 2,   3,  4, 5}}')
  })
})

describe('block node', () => {
  test('without a body', () => {
    parseSnapshot('{{#a}}{{/a}}')
  })
  test('with a variable inside', () => {
    parseSnapshot('{{#a}} {{b}} {{/a}}')
  })
  test('with a block inside', () => {
    parseSnapshot('{{#a}} {{#b}} {{/b}} {{/a}}')
  })
  test('that spans multiple lines', () => {
    parseSnapshot('{{#a}}\n  Hello world\n{{/a}}')
  })
  test('with a head and tail that span multiple lines', () => {
    parseSnapshot(`
      foo
      {{
        #a
      }}
        Hello world
      {{
        /a
      }}
      bar
    `)
  })
})

test('unclosed node', () => {
  parseThrows('{{')
  parseThrows('{{a')
  parseThrows('{{a}')
})

test('unclosed block', () => {
  parseThrows('{{#a}}')
})

test('parent block closed before child', () => {
  parseThrows('{{#a}} {{#b}} {{/a}}')
})

// Escaping is not yet supported.
xtest('escaped left bracket', () => {
  parseSnapshot('\\{{')
})
