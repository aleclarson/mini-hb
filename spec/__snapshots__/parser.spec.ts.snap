// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`block node that spans multiple lines 1`] = `
Array [
  Object {
    "body": "Hello world",
    "head": Object {
      "args": undefined,
      "end": 6,
      "line": 1,
      "name": "a",
      "start": 0,
      "type": "#",
    },
    "parent": null,
    "tail": Object {
      "args": undefined,
      "end": 27,
      "line": 3,
      "name": "a",
      "start": 21,
      "type": "/",
    },
  },
]
`;

exports[`block node with a block inside 1`] = `
Array [
  Object {
    "body": " {{#b}} {{/b}} ",
    "head": Object {
      "args": undefined,
      "end": 6,
      "line": 1,
      "name": "a",
      "start": 0,
      "type": "#",
    },
    "parent": null,
    "tail": Object {
      "args": undefined,
      "end": 27,
      "line": 1,
      "name": "a",
      "start": 21,
      "type": "/",
    },
  },
]
`;

exports[`block node with a head and tail that span multiple lines 1`] = `
Array [
  Object {
    "body": "      Hello world",
    "head": Object {
      "args": undefined,
      "end": 39,
      "line": 3,
      "name": "a",
      "start": 17,
      "type": "#",
    },
    "parent": null,
    "tail": Object {
      "args": undefined,
      "end": 88,
      "line": 7,
      "name": "a",
      "start": 66,
      "type": "/",
    },
  },
]
`;

exports[`block node with a variable inside 1`] = `
Array [
  Object {
    "body": " {{b}} ",
    "head": Object {
      "args": undefined,
      "end": 6,
      "line": 1,
      "name": "a",
      "start": 0,
      "type": "#",
    },
    "parent": null,
    "tail": Object {
      "args": undefined,
      "end": 19,
      "line": 1,
      "name": "a",
      "start": 13,
      "type": "/",
    },
  },
]
`;

exports[`block node without a body 1`] = `
Array [
  Object {
    "body": "",
    "head": Object {
      "args": undefined,
      "end": 6,
      "line": 1,
      "name": "a",
      "start": 0,
      "type": "#",
    },
    "parent": null,
    "tail": Object {
      "args": undefined,
      "end": 12,
      "line": 1,
      "name": "a",
      "start": 6,
      "type": "/",
    },
  },
]
`;

exports[`empty node 1`] = `"Invalid variable name \\"\\" on line 1"`;

exports[`function call with creative spacing 1`] = `
Array [
  Object {
    "args": Array [
      "1",
      "2",
      "3",
      "4",
      "5",
    ],
    "end": 25,
    "line": 1,
    "name": "fn",
    "start": 0,
    "type": undefined,
  },
]
`;

exports[`function call with many arguments 1`] = `
Array [
  Object {
    "args": Array [
      "z0",
      "z1",
      "z2",
      "z3",
      "z4",
      "z5",
      "z6",
      "z7",
      "z8",
      "z9",
    ],
    "end": 47,
    "line": 1,
    "name": "fn",
    "start": 0,
    "type": undefined,
  },
]
`;

exports[`function call with no argument spacing 1`] = `
Array [
  Object {
    "args": Array [
      "1",
      "2",
      "a",
      "b",
      "null",
    ],
    "end": 19,
    "line": 1,
    "name": "fn",
    "start": 0,
    "type": undefined,
  },
]
`;

exports[`function call with one argument 1`] = `
Array [
  Object {
    "args": Array [
      "z",
    ],
    "end": 10,
    "line": 1,
    "name": "fn",
    "start": 0,
    "type": undefined,
  },
]
`;

exports[`function call with primitive arguments 1`] = `
Array [
  Object {
    "args": Array [
      "0",
      "-0.5",
      "1",
      "true",
      "false",
      "null",
    ],
    "end": 38,
    "line": 1,
    "name": "fn",
    "start": 0,
    "type": undefined,
  },
]
`;

exports[`line breaks 1`] = `
Array [
  Object {
    "args": undefined,
    "end": 9,
    "line": 3,
    "name": "a",
    "start": 4,
    "type": undefined,
  },
  Object {
    "args": undefined,
    "end": 17,
    "line": 5,
    "name": "b",
    "start": 12,
    "type": undefined,
  },
]
`;

exports[`parent block closed before child 1`] = `"Unexpected closing node on line 1"`;

exports[`unclosed block 1`] = `"Unclosed block on line 1"`;

exports[`unclosed node 1`] = `"Unclosed node on line 1"`;

exports[`unclosed node 2`] = `"Unclosed node on line 1"`;

exports[`unclosed node 3`] = `"Unclosed node on line 1"`;

exports[`variable names can use dot-notation 1`] = `
Array [
  Object {
    "args": undefined,
    "end": 13,
    "line": 1,
    "name": "a.b.c.d",
    "start": 0,
    "type": undefined,
  },
]
`;

exports[`variable names cannot be a number 1`] = `"Invalid variable name \\"12345\\" on line 1"`;

exports[`variable names cannot be a number 2`] = `"Invalid variable name \\"12.34\\" on line 1"`;

exports[`variable names cannot contain dashes 1`] = `"Invalid variable name \\"a--b-c\\" on line 1"`;

exports[`variable names cannot start with a number 1`] = `"Invalid variable name \\"0z\\" on line 1"`;

exports[`variable names cannot use bracket-notation (yet?) 1`] = `"Invalid variable name \\"a[b[c]][d]\\" on line 1"`;

exports[`variable names follow Javascript rules exactly 1`] = `
Array [
  Object {
    "args": undefined,
    "end": 28,
    "line": 1,
    "name": "__aAoOzZ_$0123456789$$",
    "start": 0,
    "type": undefined,
  },
]
`;

exports[`variable node that spans multiple lines 1`] = `
Array [
  Object {
    "args": undefined,
    "end": 38,
    "line": 3,
    "name": "a",
    "start": 17,
    "type": undefined,
  },
]
`;

exports[`variable node with common padding 1`] = `
Array [
  Object {
    "args": undefined,
    "end": 5,
    "line": 1,
    "name": "a",
    "start": 0,
    "type": undefined,
  },
]
`;

exports[`variable node with common padding 2`] = `
Array [
  Object {
    "args": undefined,
    "end": 7,
    "line": 1,
    "name": "a",
    "start": 0,
    "type": undefined,
  },
]
`;

exports[`variable node with creative padding 1`] = `
Array [
  Object {
    "args": undefined,
    "end": 6,
    "line": 1,
    "name": "a",
    "start": 0,
    "type": undefined,
  },
]
`;

exports[`variable node with creative padding 2`] = `
Array [
  Object {
    "args": undefined,
    "end": 11,
    "line": 1,
    "name": "a",
    "start": 0,
    "type": undefined,
  },
]
`;

exports[`variable node with spacing around 1`] = `
Array [
  Object {
    "args": undefined,
    "end": 8,
    "line": 1,
    "name": "a",
    "start": 2,
    "type": undefined,
  },
]
`;
